File: reference/src/security/docker-sandbox.ts
/**
 * Docker container sandbox provider (stub).
 * Implements the Sandbox interface using Docker container isolation.
 * This is a stub that shells out to `docker run` — production deployments
 * would use the Docker Engine API directly.
 */

import type { Sandbox, SandboxConstraints, SandboxStatus } from './interfaces.js';

/** Function that executes a shell command and returns stdout. */
export type ShellExec = (command: string) => Promise<string>;

export interface DockerSandboxConfig {
  /** Docker image to use for sandboxes. */
  image: string;
  /** Optional Docker network to attach (defaults to 'none'). */
  network?: string;
}

/**
 * Create a Docker-backed sandbox provider.
 *
 * @param exec - Shell execution function (e.g., wrapping child_process.execSync).
 * @param config - Docker configuration.
 */
export function createDockerSandbox(exec: ShellExec, config: DockerSandboxConfig): Sandbox {
  const containers = new Map<string, string>(); // sandboxId → containerId

  return {
    async isolate(taskId: string, constraints: SandboxConstraints): Promise<string> {
      const sandboxId = `docker-${taskId}-${Date.now()}`;
      const network = constraints.networkPolicy === 'none' ? 'none' : (config.network ?? 'bridge');
      const memoryFlag = `--memory=${constraints.maxMemoryMb}m`;
      const cpuFlag = `--cpus=${(constraints.maxCpuPercent / 100).toFixed(2)}`;
      const networkFlag = `--network=${network}`;

      const cmd = `docker run -d ${memoryFlag} ${cpuFlag} ${networkFlag} --name ${sandboxId} ${config.image} sleep ${Math.ceil(constraints.timeoutMs / 1000)}`;
      const containerId = (await exec(cmd)).trim();
      containers.set(sandboxId, containerId);
      return sandboxId;
    },

    async destroy(sandboxId: string): Promise<void> {
      const containerId = containers.get(sandboxId);
      if (!containerId) {
        throw new Error(`Sandbox "${sandboxId}" not found`);
      }
      await exec(`docker rm -f ${containerId}`);
      containers.delete(sandboxId);
    },

    async getStatus(sandboxId: string): Promise<SandboxStatus> {
      const containerId = containers.get(sandboxId);
      if (!containerId) {
        throw new Error(`Sandbox "${sandboxId}" not found`);
      }
      const state = (await exec(`docker inspect -f '{{.State.Status}}' ${containerId}`)).trim();
      switch (state) {
        case 'running':
          return 'running';
        case 'created':
          return 'idle';
        case 'exited':
        case 'dead':
        case 'removing':
          return 'terminated';
        default:
          return 'error';
      }
    },
  };
}

File: sdk-python/src/ai_sdlc/security/docker_sandbox.py
"""Docker container sandbox provider."""

from __future__ import annotations

import math
from collections.abc import Callable, Coroutine
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .interfaces import SandboxConstraints, SandboxStatus

ShellExec = Callable[[str], Coroutine[object, object, str]]


@dataclass
class DockerSandboxConfig:
    image: str
    network: str | None = None


class _DockerSandbox:
    def __init__(self, exec_fn: ShellExec, config: DockerSandboxConfig) -> None:
        self._exec = exec_fn
        self._config = config
        self._containers: dict[str, str] = {}

    async def isolate(
        self, task_id: str, constraints: SandboxConstraints,
    ) -> str:
        import time

        sandbox_id = f"docker-{task_id}-{int(time.time() * 1000)}"
        network = (
            "none" if constraints.network_policy == "none"
            else (self._config.network or "bridge")
        )
        mem = f"--memory={constraints.max_memory_mb}m"
        cpu = f"--cpus={constraints.max_cpu_percent / 100:.2f}"
        net = f"--network={network}"
        timeout_s = math.ceil(constraints.timeout_ms / 1000)

        cmd = (
            f"docker run -d {mem} {cpu} {net} "
            f"--name {sandbox_id} {self._config.image} "
            f"sleep {timeout_s}"
        )
        container_id = (await self._exec(cmd)).strip()
        self._containers[sandbox_id] = container_id
        return sandbox_id

    async def destroy(self, sandbox_id: str) -> None:
        container_id = self._containers.get(sandbox_id)
        if not container_id:
            raise KeyError(f'Sandbox "{sandbox_id}" not found')
        await self._exec(f"docker rm -f {container_id}")
        del self._containers[sandbox_id]

    async def get_status(self, sandbox_id: str) -> SandboxStatus:
        container_id = self._containers.get(sandbox_id)
        if not container_id:
            raise KeyError(f'Sandbox "{sandbox_id}" not found')
        state = (
            await self._exec(
                f"docker inspect -f '{{{{.State.Status}}}}' {container_id}",
            )
        ).strip()
        state_map: dict[str, SandboxStatus] = {
            "running": "running",
            "created": "idle",
            "exited": "terminated",
            "dead": "terminated",
            "removing": "terminated",
        }
        return state_map.get(state, "error")


def create_docker_sandbox(
    exec_fn: ShellExec, config: DockerSandboxConfig,
) -> _DockerSandbox:
    """Create a Docker-backed sandbox provider."""
    return _DockerSandbox(exec_fn, config)
